scipy/optimize/_qap.py

```diff
@@ -0,0 +1,786 @@
+
+    maxiter = operator.index(maxiter)
+
+    # ValueError check
+    A, B, partial_match = _common_input_validation(A, B, partial_match)
+
+    msg = None
+    if (A < 0).any() or (B < 0).any():
+        msg = "`A` and `B` matrices must contain only non-negative elements."
+    elif isinstance(init_J, str) and init_J not in {'barycenter'}:
+        msg = "Invalid 'init_J' parameter string"
+    elif init_weight is not None and (init_weight < 0 or init_weight > 1):
+        msg = "'init_weight' must be strictly between zero and one"
+    elif maxiter <= 0:
+        msg = "'maxiter' must be a positive integer"
+    elif tol <= 0:
+        msg = "'tol' must be a positive float"
+    if msg is not None:
+        raise ValueError(msg)
+
+    rng = check_random_state(rng)
+    n = A.shape[0]  # number of vertices in graphs
+    n_seeds = partial_match.shape[0]  # number of seeds
+    n_unseed = n - n_seeds
+
+    obj_func_scalar = 1
+    if maximize:
+        obj_func_scalar = -1
+
+    seed_dist_c = np.setdiff1d(range(n), partial_match[:, 1])
+    if shuffle_input:
+        seed_dist_c = rng.permutation(seed_dist_c)
+        # shuffle_input to avoid results from inputs that were already matched
+
+    seed_cost_c = np.setdiff1d(range(n), partial_match[:, 0])
+    perm_A = np.concatenate([partial_match[:, 0],
+                             seed_cost_c], axis=None).astype(int)
+    perm_B = np.concatenate([partial_match[:, 1],
+                             seed_dist_c], axis=None).astype(int)
+
+    A = A[np.ix_(perm_A, perm_A)]
+    B = B[np.ix_(perm_B, perm_B)]
+
+    # definitions according to Seeded Graph Matching [2].
+    A11 = A[:n_seeds, :n_seeds]
+    A12 = A[:n_seeds, n_seeds:]
+    A21 = A[n_seeds:, :n_seeds]
+    A22 = A[n_seeds:, n_seeds:]
+    B11 = B[:n_seeds, :n_seeds]
+    B12 = B[:n_seeds, n_seeds:]
+    B21 = B[n_seeds:, :n_seeds]
+    B22 = B[n_seeds:, n_seeds:]
+
+    # setting initialization matrix
+    if isinstance(init_J, str) and init_J == 'barycenter':
+        J = np.ones((n_unseed, n_unseed)) / float(n_unseed)
+    else:
+        _check_init_input(init_J, n_unseed)
+        J = init_J
+
+    if init_weight is not None:
+        # generate a nxn matrix where each entry is a random number [0, 1]
+        # would use rand, but Generators don't have it
+        # would use random, but old mtrand.RandomStates don't have it
+        K = rng.uniform(size=(n_unseed, n_unseed))
+        # Sinkhorn balancing
+        K = _doubly_stochastic(K)
+        # initialize J, a doubly stochastic barycenter
+        P = J * init_weight + (1 - init_weight) * K
+    else:
+        P = J
+    const_sum = A21 @ B21.T + A12.T @ B12
+    grad_P = np.inf  # gradient of P
+    n_iter = 0  # number of FW iterations
+
+    # OPTIMIZATION WHILE LOOP BEGINS
+    while grad_P > tol and n_iter < maxiter:
+        # computing the gradient of f(P) = -tr(APB^tP^t)
+        delta_f = (const_sum + A22 @ P @ B22.T + A22.T @ P @ B22)
```

Comment:
"delta" is a misleading variable name for the gradient. The FAQ paper uses nabla, but `grad_fp` would also do.